Add non-allocating helper function for torch.linalg.qr

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57317 Add cuSOLVER path for torch.linalg.lstsq
* #57316 Add CUDA support for torch.ormqr
* #57315 Port CPU torch.ormqr to ATen
* #57314 Fix torch.ormqr for non Fortran-contiguous inputs
* #56257 Fix MAGMA qr for empty batched inputs
* #56256 Add cuSOLVER path for torch.linalg.qr
* #56255 Remove size arguments for internal orgqr and geqrf calls
* **#56254 Add non-allocating helper function for torch.linalg.qr**

Differential Revision: [D27960151](https://our.internmc.facebook.com/intern/diff/D27960151)

IvanYashchuk committed:
Add non-allocating helper function for torch.linalg.qr

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"




[ghstack-poisoned]

mruberry commented:
@Lezcano would you sanity check this?

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"


Differential Revision: [D27960151](https://our.internmc.facebook.com/intern/diff/D27960151)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"


Differential Revision: [D27960151](https://our.internmc.facebook.com/intern/diff/D27960151)

[ghstack-poisoned]

antocuni commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1693,55 +1693,119 @@ std::tuple<Tensor, Tensor> geqrf(const Tensor& input) {
+    TORCH_INTERNAL_ASSERT(Q.sizes().equals(expected_Q_shape));
+
+    // Q tensor must be in batched column major order (Fortran contiguous)
+    TORCH_INTERNAL_ASSERT(Q.transpose(-2, -1).is_contiguous());
```
this has nothing to do with this PR, it's just a random idea which came to my mind while reading this.
Should we have something like `is_fortran_contiguous()`? And maybe even introduce `at::MemoryFormat::FortranContiguous` which would be useful when creating tensors.
I have seen similar patterns a bit everywhere in linalg code, and having first-class support for fortran-style memory would probably simplify a lot of code

IvanYashchuk replied:
In linalg code we need to have only that the last two dimensions to be Fortran contiguous and all other are normal C contiguous. Real Fortran-style memory would be to have the matrix in the first two dimensions and batches in the end, it's not useful to have.

antocuni reviewed (commented):
LGTM.
If I understand it correctly, this PR by itself is mostly a no-op because it introduces `_linalg_qr_out_helper` but it never uses it apart for re-implementing the old `_linalg_qr_helper`, but I assume it will be used by subsequent PRs in the ghstack

lezcano reviewed (commented):
It looks very good. I just found one part that was a bit difficult to follow. I made a proposal for a simpler way to handle that, which I think should be save one allocation.

If this sounds reasonable, the handling of the "unpacking" of the tensors after the `geqrf` call should be changed as well.

mruberry commented:
@Lezcano would you take a look at this PR?

IvanYashchuk commented:
> If I understand it correctly, this PR by itself is mostly a no-op because it introduces _linalg_qr_out_helper but it never uses it apart for re-implementing the old _linalg_qr_helper, but I assume it will be used by subsequent PRs in the ghstack

The previous helper was allocating memory on the go, with narrowing and resizes, which is not a good strategy to do. Here we first allocate the memory of the correct size and then fill it with the result.

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1693,55 +1693,119 @@ std::tuple<Tensor, Tensor> geqrf(const Tensor& input) {
+  // geqrf requires m x n workspace input that is modified in-place
+  // if m > n and reduced==true we use Q tensor for storing the result of geqrf operation
+  // otherwise R tensor is used
+  Tensor QR = R;
```
> the current implementation allocates a new tensor whenever compute_q == false, which is not necessary

Right, I'll fix that.

IvanYashchuk replied:
I changed the code to use your conditions.

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"


Differential Revision: [D27960151](https://our.internmc.facebook.com/intern/diff/D27960151)

[ghstack-poisoned]

lezcano reviewed (approved):
LGTM!

IvanYashchuk committed:
Update on "Add non-allocating helper function for torch.linalg.qr"


Differential Revision: [D27960151](https://our.internmc.facebook.com/intern/diff/D27960151)

[ghstack-poisoned]

mruberry reviewed (approved):
Stamped
